src/components/structures/ScrollPanel.js

```diff
@@ -128,13 +153,76 @@ module.exports = React.createClass({
     checkFillState: function() {
         var sn = this._getScrollNode();
 
+        // if there is less than a screenful of messages above or below the
+        // viewport, try to get some more messages.
+        //
+        // scrollTop is the number of pixels between the top of the content and
+        //     the top of the viewport.
+        //
+        // scrollHeight is the total height of the content.
+        //
+        // clientHeight is the height of the viewport (excluding borders,
+        // margins, and scrollbars).
+        //
+        //
+        //   .---------.          -                 -
+        //   |         |          |  scrollTop      |
+        // .-+---------+-.    -   -                 |
+        // | |         | |    |                     |
+        // | |         | |    |  clientHeight       | scrollHeight
+        // | |         | |    |                     |
+        // `-+---------+-'    -                     |
+        //   |         |                            |
+        //   |         |                            |
+        //   `---------'                            -
+        //
+
         if (sn.scrollTop < sn.clientHeight) {
-            // there's less than a screenful of messages left - try to get some
-            // more messages.
-            this.props.onFillRequest(true);
+            // need to back-fill
+            this._maybeFill(true);
+        }
+        if (sn.scrollTop > sn.scrollHeight - sn.clientHeight * 2) {
```

Comment:
That is what we're doing. Start with your expression:

    (sn.scrollHeight - sn.clientHeight - sn.scrollTop) < sn.clientHeight

Subtract `sn.clientHeight` from both sides:

    (sn.scrollHeight - sn.clientHeight * 2 - sn.scrollTop) < 0

Add `sn.scrollTop`:

    (sn.scrollHeight - sn.clientHeight * 2) < sn.scrollTop

Reverse the expression:

    sn.scrollTop > sn.scrollHeight - sn.clientHeight * 2

As an alternative way of thinking about it: look at the ascii art - I want to make sure that there are at least two screenfuls below scrollTop - one for the actual viewport and one for the 'bottom height'. So I want scrollTop to be at most scrollHeight - clientHeight * 2.